File: python/zpy/src/zingg_py/llm_labeller.py
"""LLM-backed labeller — port of ``LlmLabeller.scala``.

Calls an OpenAI-compatible chat-completions endpoint, one request per pair, and
parses a single 1/0/-1 verdict. Pairs the model can't decide (-1, unparseable,
or HTTP error) are skipped. Uses stdlib ``urllib`` — no extra dependency.

Env (constructor args override):
  ZINGG_AI_ENDPOINT  full URL  (default http://localhost:11434/v1/chat/completions)
                     — or set OPENAI_BASE (e.g. http://127.0.0.1:8080/v1) and the
                       /chat/completions suffix is appended.
  ZINGG_AI_MODEL     default llama3
  ZINGG_AI_KEY       sent as `Authorization: Bearer …` if set
"""

from __future__ import annotations

import json
import os
import re
import socket
import urllib.parse
import urllib.request

from .labeller import Decision, RowLabeller
from .schema import ZinggConf

_LAST_SIGNED_DIGIT = re.compile(r"(?<![0-9])(-?[01])(?![0-9])")


def _default_endpoint() -> str:
    if "ZINGG_AI_ENDPOINT" in os.environ:
        return os.environ["ZINGG_AI_ENDPOINT"]
    base = os.environ.get("OPENAI_BASE")
    if base:
        return base.rstrip("/") + "/chat/completions"
    return "http://localhost:11434/v1/chat/completions"


def endpoint_reachable(url: str | None = None, timeout: float = 1.0) -> bool:
    """Cheap TCP reachability check for the configured LLM endpoint."""
    parsed = urllib.parse.urlparse(url or _default_endpoint())
    host = parsed.hostname
    if host is None:
        return False
    port = parsed.port or (443 if parsed.scheme == "https" else 80)
    try:
        with socket.create_connection((host, port), timeout=timeout):
            return True
    except OSError:
        return False


def first_signed_digit(s: str):
    """LAST standalone -1/0/1 in the text (reasoning models end with the answer)."""
    matches = _LAST_SIGNED_DIGIT.findall(s)
    return int(matches[-1]) if matches else None


def parse_label(body: str):
    """Extract the verdict from an OpenAI chat-completions response."""
    root = json.loads(body)
    choices = root.get("choices")
    if not isinstance(choices, list) or not choices:
        return None
    msg = choices[0].get("message", {})
    for key in ("content", "reasoning_content", "reasoning"):
        text = msg.get(key) or ""
        if text.strip():
            return first_signed_digit(text)
    return None


class LlmLabeller(RowLabeller):
    def __init__(self, endpoint: str | None = None, model: str | None = None,
                 api_key: str | None = None, timeout: float = 60.0):
        self.endpoint = endpoint or _default_endpoint()
        self.model = model or os.environ.get("ZINGG_AI_MODEL", "llama3")
        self.api_key = api_key if api_key is not None else os.environ.get("ZINGG_AI_KEY")
        self.timeout = timeout

    def decide(self, r, cfg: ZinggConf) -> Decision:
        v = self._ask_one(r, cfg)
        if v == 1:
            return Decision.Match
        if v == 0:
            return Decision.NonMatch
        return Decision.Unknown

    def _ask_one(self, r, cfg: ZinggConf):
        try:
            return parse_label(self._call_api(self._build_prompt(r, cfg)))
        except Exception:
            return None

    def _build_prompt(self, r, cfg: ZinggConf) -> str:
        def val(prefix, name):
            v = r.get(f"{prefix}{name}")
            return "" if v is None else str(v)

        rendered = "\n".join(
            f"  {f.name:<20} | A: {val(ZinggConf.LEFT_PREFIX, f.name)}\n"
            f"  {'':<20}  | B: {val(ZinggConf.RIGHT_PREFIX, f.name)}"
            for f in cfg.fields
        )
        return (
            "You are deciding whether two records refer to the same real-world entity.\n"
            "Answer with EXACTLY one digit on a line by itself, no other text:\n"
            "  1  = same entity (match)\n"
            "  0  = different entities (non-match)\n"
            " -1  = cannot tell from the information given\n\n"
            f"Record pair:\n{rendered}\n\nAnswer:"
        )

    def _call_api(self, prompt: str) -> str:
        payload = {
            "model": self.model,
            "temperature": 0.0,
            "messages": [{"role": "user", "content": prompt}],
        }
        data = json.dumps(payload).encode("utf-8")
        req = urllib.request.Request(self.endpoint, data=data, method="POST")
        req.add_header("Content-Type", "application/json")
        if self.api_key:
            req.add_header("Authorization", f"Bearer {self.api_key}")
        with urllib.request.urlopen(req, timeout=self.timeout) as resp:
            if resp.status != 200:
                raise RuntimeError(f"LLM HTTP {resp.status}")
            return resp.read().decode("utf-8")

File: zscala/src/main/scala/zingg/zs/LlmLabeller.scala
package zingg.zs

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.node.ObjectNode

import java.net.URI
import java.net.http.{HttpClient, HttpRequest, HttpResponse}
import java.nio.charset.StandardCharsets
import java.time.Duration

import org.apache.spark.sql.Row

/** LLM-backed labeller for OpenAI-compatible chat completion endpoints.
  *
  * Defaults to a local Ollama instance; override via env or constructor:
  *
  *   ZINGG_AI_ENDPOINT   default http://localhost:11434/v1/chat/completions
  *   ZINGG_AI_MODEL      default llama3
  *   ZINGG_AI_KEY        default unset (sent as `Authorization: Bearer …` if set)
  *
  * The prompt asks the model to answer with a single digit (1/0/-1). Pairs
  * the model can't decide on (-1, unparseable, or HTTP error) are skipped
  * — they're dropped from the returned DataFrame, not labelled.
  *
  * One HTTP call per pair. Active-learning batches are small (~30 pairs)
  * so latency rarely matters; if you do need throughput, wrap with a
  * thread pool — this class is thread-safe modulo the underlying HttpClient.
  */
final class LlmLabeller(
    endpoint: String = sys.env.getOrElse("ZINGG_AI_ENDPOINT", "http://localhost:11434/v1/chat/completions"),
    model:    String = sys.env.getOrElse("ZINGG_AI_MODEL",    "llama3"),
    apiKey:   Option[String] = sys.env.get("ZINGG_AI_KEY"),
    requestTimeout: Duration = Duration.ofSeconds(60)
) extends RowLabeller {

  @transient private lazy val mapper = new ObjectMapper()
  @transient private lazy val http   = HttpClient.newBuilder()
    .connectTimeout(Duration.ofSeconds(10))
    .build()

  def decide(r: Row, cfg: ZinggConf): RowLabeller.Decision =
    askOne(r, cfg) match {
      case Some(1) => RowLabeller.Match
      case Some(0) => RowLabeller.NonMatch
      case _       => RowLabeller.Unknown  // unsure / error – defer
    }

  // Visible for testing.
  private[zs] def askOne(r: Row, cfg: ZinggConf): Option[Int] = {
    val prompt = buildPrompt(r, cfg)
    try {
      val body = callApi(prompt)
      parseLabel(body)
    } catch {
      case _: Throwable => None
    }
  }

  private def buildPrompt(r: Row, cfg: ZinggConf): String = {
    val left  = cfg.fields.map(f => f.name -> Option(r.getAs[Any](
      s"${ZinggConf.LeftPrefix}${f.name}")).map(_.toString).getOrElse("")).toMap
    val right = cfg.fields.map(f => f.name -> Option(r.getAs[Any](
      s"${ZinggConf.RightPrefix}${f.name}")).map(_.toString).getOrElse("")).toMap
    val rendered = cfg.fields.map { f =>
      f"  ${f.name}%-20s | A: ${left(f.name)}%n  ${" " * 20}  | B: ${right(f.name)}"
    }.mkString("\n")
    s"""You are deciding whether two records refer to the same real-world entity.
       |Answer with EXACTLY one digit on a line by itself, no other text:
       |  1  = same entity (match)
       |  0  = different entities (non-match)
       | -1  = cannot tell from the information given
       |
       |Record pair:
       |$rendered
       |
       |Answer:""".stripMargin
  }

  private def callApi(prompt: String): String = {
    val body = mapper.createObjectNode()
    body.put("model", model)
    body.put("temperature", 0.0)
    val messages = body.putArray("messages")
    val userMsg = messages.addObject()
    userMsg.put("role", "user")
    userMsg.put("content", prompt)

    val builder = HttpRequest.newBuilder()
      .uri(URI.create(endpoint))
      .timeout(requestTimeout)
      .header("Content-Type", "application/json")
      .POST(HttpRequest.BodyPublishers.ofString(mapper.writeValueAsString(body),
                                                 StandardCharsets.UTF_8))
    apiKey.foreach(k => builder.header("Authorization", s"Bearer $k"))

    val response = http.send(builder.build(), HttpResponse.BodyHandlers.ofString())
    if (response.statusCode() != 200)
      throw new RuntimeException(s"LLM HTTP ${response.statusCode()}: ${response.body()}")
    response.body()
  }

  /** Extract the integer label from an OpenAI chat-completions response.
    * Handles `message.content` and the `reasoning_content` field used by
    * Qwen3 / DeepSeek when their normal content is empty.
    */
  private[zs] def parseLabel(json: String): Option[Int] = {
    val root = mapper.readTree(json)
    val choices = root.path("choices")
    if (!choices.isArray || choices.size() == 0) return None
    val msg = choices.get(0).path("message")
    val text = Seq("content", "reasoning_content", "reasoning")
      .map(msg.path).map(n =>
        if (n.isMissingNode || n.isNull) "" else n.asText("")
      ).find(_.trim.nonEmpty).getOrElse("")
    LlmLabeller.firstSignedDigit(text)
  }
}

object LlmLabeller {
  /** Find the LAST standalone occurrence of `-1`, `0`, or `1` in the text —
    * standalone meaning not part of a larger number like `1234` or `2024`.
    *
    * "Last" because reasoning models put the final answer at the end after
    * chain-of-thought. CVE ids and other digit-runs embedded in the text
    * are skipped via the lookahead/lookbehind constraints.
    */
  private[zs] def firstSignedDigit(s: String): Option[Int] = {
    val re = "(?<![0-9])(-?[01])(?![0-9])".r
    re.findAllMatchIn(s).toList.lastOption.map(_.group(1).toInt)
  }
}
